File: valid_perfect_square.js
var isPerfectSquare = function(num) {
  if (num === 0 || num === 1) return true;

  var left = 1;
  var right = num;
  var mid;

  while (left + 1 < right) {
    mid = Math.floor(left + (right - left) / 2);

    if (mid * mid === num) {
      return true;
    } else if (mid * mid < num) {
      left = mid;
    } else {
      right = mid;
    }
  }

  if (left  * left  === num) return left;
  if (right * right === num) return right;

  return false;
};

console.log(isPerfectSquare(4)) // true
console.log(isPerfectSquare(9)) // true
console.log(isPerfectSquare(5)) // false

console.log(isPerfectSquare(81)) // true
console.log(isPerfectSquare(75)) // false
console.log(isPerfectSquare(16)) // true
console.log(isPerfectSquare(14)) // false
console.log(isPerfectSquare(1)) // true
console.log(isPerfectSquare(0)) // true

File: binary_search/valid_perfect_square.rb
# Given a positive integer num, write a function which returns
# True if num is a perfect square else False.

# Note: Do not use any built-in library function such as sqrt.

# Example 1:

# Input: 16
# Returns: True
# Example 2:

# Input: 14
# Returns: False

# perfect square means there is some n such that num = n * n
# start at midpoint, see if mid * mid == num
# if it is return True
# else keep going

# end return false

# what if number is 0 or 1?

# Binary Search
# T: O(logN), S: O(1)

# @param {Integer} num
# @return {Boolean}
def is_perfect_square num
  return true if num.zero?

  low  = 0
  high = num

  while low + 1 < high
    mid  = low + (high - low) / 2

    if mid * mid == num
      return true
    elsif mid * mid > num # look in left half
      high = mid
    else # look in right half
      low = mid
    end
  end

  return true if low  * low  == num
  return true if high * high == num
  false
end

puts is_perfect_square 16 # true
puts is_perfect_square 14 # false

